### PR TITLE
[aDAG] Support multi-read of the same shm channel

### LINKED ISSUE #47041: [core][aDAG] Hangs if same shared memory channel is read twice by one actor
### What happened + What you expected to happen

```
driver -> actor1.op ...
       |
       -> actor1.op ...
```

* This doesn't work. We should pass information about how many times the `MutableObject` should be read per actor from Python to C++ to avoid hanging.
  ```python
  with InputNode() as inp:
      d00 = worker_0.forward.bind(inp)  # worker_0 FWD
      d01 = worker_0.forward.bind(inp)  # worker_0 FWD
      d02 = worker_0.forward.bind(inp)  # worker_0 FWD
  ```

* Workaround: This works. Read the shared memory once, and then use `IntraProcessChannel` to transfer data between `read_input` and `forward`. In addition, this solution is more efficient. See #46431 for more details.
  ```python
  with InputNode() as inp:
      w0_input = worker_0.read_input.bind(inp)
      d00 = worker_0.forward.bind(w0_input, 0)  # worker_0 FWD
      d01 = worker_0.forward.bind(w0_input, 1)  # worker_0 FWD
      d02 = worker_0.forward.bind(w0_input, 2)  # worker_0 FWD
  ```

### Versions / Dependencies

nightly

### Reproduction script

see above

### Issue Severity

None

### PR DESCRIPTION
**Pull request by ruisearch42**
## Why are these changes needed?

If the same method of the same actor is bound to the same node (i.e., reads from the same shared memory channel), aDAG execution hangs. This PR adds support to this case by caching results read from the channel.

## Related issue number

Closes #47041
## Checks

- [ ] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
    - [ ] I've added any new APIs to the API Reference. For example, if I added a 
           method in Tune, I've added it in `doc/source/tune/api/` under the 
           corresponding `.rst` file.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

**Review by rkooo567 — COMMENTED** [review]
QQ: is this approach working for this kind of case? (I believe it will not work?)

```python3

with InputNode as inp:
    out = s1.fwd(inp)
    dag = s2.fwd(inp, out)
```

Also have some concerns that deepcopy on the first result can affect perf negatively. 

I actually wonder if it is viable approach to just not allow the additional write until all downstream tasks finish reading (I think with buffering input PR, it may work well?)

**Review by rkooo567 — COMMENTED** [review]
discussed offline. deep/shallow copy overhead is pretty big, so we will make some assumptions that input won't be changed (which is not very correct). we can allow copy inputs with some sort of flags

File: python/ray/dag/compiled_dag_node.py [hunk 1]
```diff
@@ -302,26 +343,31 @@ def __init__(
         # its index in input_channels.
         input_channel_to_idx: dict[ChannelInterface, int] = {}
 
-        for arg in resolved_args:
+        for tuple in resolved_args:
```
Comment (rkooo567) [inline python/ray/dag/compiled_dag_node.py:346]:
```suggestion
        for arg_and_use_cached in resolved_args:
```

File: python/ray/dag/compiled_dag_node.py [hunk 2]
```diff
@@ -302,26 +343,31 @@ def __init__(
...
-                    # The same channel was added before, so reuse the index.
-                    channel_idx = input_channel_to_idx[channel]
+                if use_cached:
+                    task_input = _ExecutableTaskInput(arg, None, True)
                 else:
-                    # Add a new channel to the list of input channels.
-                    self.input_channels.append(channel)
-                    channel_idx = len(self.input_channels) - 1
-                    input_channel_to_idx[channel] = channel_idx
+                    if channel in input_channel_to_idx:
```
Comment (rkooo567) [inline python/ray/dag/compiled_dag_node.py:358]: is this related to this PR?

File: python/ray/dag/compiled_dag_node.py [hunk 3]
```diff
@@ -464,14 +516,15 @@ def exec_operation(self, class_handle, op_type: _DAGNodeOperationType) -> bool:
             class_handle: The handle of the class to which the actor belongs.
             op_type: The type of the operation. Possible types are READ,
                 COMPUTE, and WRITE.
+            cache: A cache for input channel results.
```
Comment (rkooo567) [inline python/ray/dag/compiled_dag_node.py:519]:
```suggestion
            input_cache: A cache to for input channel results. It is in-place updated when a new input is found
```

File: python/ray/dag/compiled_dag_node.py [hunk 4]
```diff
@@ -426,7 +473,7 @@ def _compute(self, class_handle) -> bool:
 
         resolved_inputs = []
         for task_input in self.task_inputs:
-            resolved_inputs.append(task_input.resolve(input_data))
+            resolved_inputs.append(task_input.resolve(input_data, cache))
```
Comment (rkooo567) [inline python/ray/dag/compiled_dag_node.py:476]: can you rename cache -> input_result_cache

File: python/ray/tests/test_channel.py [hunk 5]
```diff
@@ -922,11 +922,11 @@ def write(self, value):
     )
     ray.get(actor1.write.remote("hello world"))
     assert ray.get(actor1.read.remote()) == "hello world"
-    assert ray.get(actor1.read.remote()) == "hello world"
```
Comment (rkooo567) [inline python/ray/tests/test_channel.py:925]: why is it removed?

File: python/ray/dag/tests/experimental/test_accelerated_dag.py [hunk 6]
```diff
@@ -267,6 +267,39 @@ def test_actor_method_bind_same_constant(ray_start_regular):
     compiled_dag.teardown()
 
 
+def test_actor_method_bind_same_arg(ray_start_regular):
```
Comment (rkooo567) [inline python/ray/dag/tests/experimental/test_accelerated_dag.py:270]: QQ: this is something intra process channel was already supposed to handle?

File: python/ray/dag/tests/experimental/test_accelerated_dag.py [hunk 7]
```diff
@@ -267,6 +267,39 @@ def test_actor_method_bind_same_constant(ray_start_regular):
...
+        output1 = actor.inc.bind(inp)
+        output2 = actor.inc.bind(inp)
+        dag = MultiOutputNode([output1, output2])
+    compiled_dag = dag.experimental_compile()
+    ref = compiled_dag.execute(1)
+    result = ray.get(ref)
+    assert result == [1, 2]
+    compiled_dag.teardown()
+
+
```
Comment (rkooo567) [inline python/ray/dag/tests/experimental/test_accelerated_dag.py:302]: can you also add this case? Want to test input is used by 2 workers + used twice by 1 worker at the same time
```
with InputNode as inp:
    out = s1.fwd(inp)
    out2 = s2.fwd(inp, out)
    out3 = s1.fwd_2(out2)
```

File: python/ray/dag/compiled_dag_node.py [hunk 8]
```diff
@@ -1036,24 +1089,13 @@ def _get_or_compile(
...
-                reader_handles_set = set()
                 for idx in task.downstream_actor_idxs:
                     reader_task = self.idx_to_task[idx]
                     assert isinstance(reader_task.dag_node, ClassMethodNode)
                     reader_handle = reader_task.dag_node._get_actor_handle()
-                    if reader_handle not in reader_handles_set:
-                        reader_and_node_list.append(
-                            (reader_handle, self._get_node_id(reader_handle))
-                        )
-                    reader_handles_set.add(reader_handle)
```
Comment (rkooo567) [inline python/ray/dag/compiled_dag_node.py:1056]: why removing this?

**Review by rkooo567 — COMMENTED** [review]
QQ: is intraprocess channel still needed after this change?

File: python/ray/dag/compiled_dag_node.py [hunk 9]
```diff
@@ -108,11 +108,16 @@ def do_exec_tasks(
             task.prepare()
 
         done = False
+        # Cache to store channel results for an actor loop iteration.
+        cache: Dict[ChannelInterface, Any] = {}
```
[the remark this answers is not in the document]
Comment (stephanie-wang) [reply python/ray/dag/compiled_dag_node.py:112]: Not sure what "schedule" means here.

File: python/ray/dag/compiled_dag_node.py [hunk 10]
```diff
@@ -235,31 +240,67 @@ class _ExecutableTaskInput:
...
+        # channel results should not be mutated by the actor methods.
+        # When the user needs to modify the channel results, they should
+        # make a copy of the channel results and modify the copy.
+        # This is the same contract as used in IntraProcessChannel.
+        # This contract is NOT enforced right now in either case.
+        # TODO(rui): introduce a flag to control the behavior:
+        # for example, by default we make a deep copy of the channel
+        # result, but the user can turn off the deep copy for performance
+        # improvements.
+        if self.use_cached:
```
[the remark this answers is not in the document]
Comment (stephanie-wang) [reply python/ray/dag/compiled_dag_node.py:290]: I agree with Sang, I don't think we need the extra `use_cached` flag?

File: python/ray/dag/compiled_dag_node.py [hunk 11]
```diff
@@ -108,11 +108,16 @@ def do_exec_tasks(
             task.prepare()
 
         done = False
+        # Cache to store channel results for an actor loop iteration.
+        cache: Dict[ChannelInterface, Any] = {}
         while True:
             if done:
                 break
+            cache.clear()
```
Comment (stephanie-wang) [inline python/ray/dag/compiled_dag_node.py:116]: Can we update / clear the cache during the DAG execution? Otherwise we may leak memory especially for GPU tensors. We can also unify this codepath with the ref counting that is currently done for intra-process channels.

**Review by stephanie-wang — COMMENTED** [review]
Looks like a good start but let's try not to introduce the new `use_cached` flag if possible (it seems to make the code more complicated) and ideally clean up the cache once we know no more tasks will need it.

One suggestion might be to have an args cache per task, instead of a cache per channel. Then we can also do a list lookup instead of dict. Like:
```python
cache: List[List[Any]]
```
The outer list index is the task idx on that actor, and the inner list is the resolved args for that task. The first time we deserialize an arg, we put it into the inner lists for all reader tasks.

File: python/ray/dag/compiled_dag_node.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (ruisearch42) [reply python/ray/dag/compiled_dag_node.py:113]: Changed with a different impl and this is removed.

File: python/ray/dag/compiled_dag_node.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (ruisearch42) [reply python/ray/dag/compiled_dag_node.py]: Removed.

File: python/ray/dag/compiled_dag_node.py [hunk 12]
```diff
@@ -271,16 +312,16 @@ class ExecutableTask:
     def __init__(
         self,
         task: "CompiledTask",
-        resolved_args: List[Any],
+        resolved_args: List[Tuple[Any, bool]],
```
[the remark this answers is not in the document]
Comment (ruisearch42) [reply python/ray/dag/compiled_dag_node.py]: Code is refactored and this is removed.

File: python/ray/dag/compiled_dag_node.py (hunk 1 above)
Comment (ruisearch42) [reply python/ray/dag/compiled_dag_node.py]: Code is refactored and this is removed.

File: python/ray/dag/compiled_dag_node.py (hunk 2 above)
Comment (ruisearch42) [reply python/ray/dag/compiled_dag_node.py]: Code is refactored and this is removed.

File: python/ray/dag/compiled_dag_node.py (hunk 3 above)
Comment (ruisearch42) [reply python/ray/dag/compiled_dag_node.py:483]: Code is refactored and this is gone.

File: python/ray/dag/compiled_dag_node.py (hunk 4 above)
Comment (ruisearch42) [reply python/ray/dag/compiled_dag_node.py:440]: Code is refactored and this is gone.

File: python/ray/dag/tests/experimental/test_accelerated_dag.py (hunk 6 above)
Comment (ruisearch42) [reply python/ray/dag/tests/experimental/test_accelerated_dag.py:270]: That's right. I will clean it up.

File: python/ray/dag/compiled_dag_node.py (hunk 8 above)
Comment (ruisearch42) [reply python/ray/dag/compiled_dag_node.py:1062]: This dedups the readers which is incorrect when computing number of readers when we support multi-reads.

File: python/ray/dag/compiled_dag_node.py (hunk 11 above)
Comment (ruisearch42) [reply python/ray/dag/compiled_dag_node.py:117]: Done.

File: python/ray/experimental/channel/cached_channel.py [hunk 13]
```diff
@@ -0,0 +1,95 @@
+import uuid
+from typing import Any, Optional
+
+from ray.experimental.channel import ChannelContext
+from ray.experimental.channel.common import ChannelInterface
+from ray.util.annotations import PublicAPI
+
+
+@PublicAPI(stability="alpha")
```
Comment (stephanie-wang) [inline python/ray/experimental/channel/cached_channel.py:9]: Remove (this is an internal API).

File: python/ray/experimental/channel/cached_channel.py [hunk 14]
```diff
@@ -0,0 +1,95 @@
...
+    avoided.
+
+    Args:
+        num_readers: The number of readers that will read from this channel. Readers
+            can be the same method of the same actor.
+    """
+
+    def __init__(
+        self,
+        num_readers,
```
Comment (stephanie-wang) [inline python/ray/experimental/channel/cached_channel.py:25]:
```suggestion
        num_readers: int,
```

File: python/ray/experimental/channel/cached_channel.py [hunk 15]
```diff
@@ -0,0 +1,95 @@
...
+    """
+    CachedChannel wraps an inner channel and caches the data read from it until
+    `num_readers` of readers have read the data. If inner channel is None, the data
+    is written to serialization context and retrieved from there. This is useful
+    when passing data within the same actor and a shared memory channel can be
+    avoided.
+
+    Args:
+        num_readers: The number of readers that will read from this channel. Readers
+            can be the same method of the same actor.
```
Comment (stephanie-wang) [inline python/ray/experimental/channel/cached_channel.py:20]: *must* instead of "can"?

Might also want to specify if this is number of tasks or number of task args, i.e. what happens if the same task reads an arg twice? Can we also add a test for that?

File: python/ray/experimental/channel/cached_channel.py [hunk 16]
```diff
@@ -0,0 +1,95 @@
...
+            return ctx.get_data(self._channel_id)
+
+        assert (
+            self._inner_channel is not None
+        ), "Cannot read from the serialization context while inner channel is None."
+        value = self._inner_channel.read(timeout)
+        # The data is already read once from the inner channel,
+        # the rest of reads will be from the channel context.
+        num_readers = self._num_readers - 1
+        ctx.set_data(self._channel_id, value, num_readers)
```
Comment (stephanie-wang) [inline python/ray/experimental/channel/cached_channel.py:78]: Don't we need to clear the data on the last read instead of setting it again? Can you make sure to check the condition (serialization context is empty after a DAG execution completes) in a unit test?

File: python/ray/dag/compiled_dag_node.py [hunk 17]
```diff
@@ -1110,34 +1114,76 @@ def _get_or_compile(
 
         # Create executable tasks for each actor
         for actor_handle, tasks in self.actor_to_tasks.items():
-            executable_tasks = []
+            # The key is the channel that the actor is reading from,
+            # the value is the number of times this channel is read
+            # by the actor.
+            actor_input_channels = defaultdict(int)
```
Comment (stephanie-wang) [inline python/ray/dag/compiled_dag_node.py:1120]:
```suggestion
            num_channel_reads = defaultdict(int)
```
Also add type hint.

File: python/ray/dag/compiled_dag_node.py [hunk 18]
```diff
@@ -1110,34 +1114,76 @@ def _get_or_compile(
...
                     )
+
+            # Step 2: create cached channels if needed
+
+            # The key is the original input channel, the value is the cached channel
+            # to use in task execution.
+            actor_cached_channels = {}
+            for channel, count in actor_input_channels.items():
+                if count > 1:
+                    actor_cached_channels[channel] = CachedChannel(count, channel)
```
Comment (stephanie-wang) [inline python/ray/dag/compiled_dag_node.py:1153]: Later code would be simpler if we overwrite self.dag_input_channel / task.output_channels (or add an actor_cached_channels entry for every channel instead of just the ones with count > 1).

File: python/ray/experimental/channel/shared_memory_channel.py [hunk 19]
```diff
@@ -518,15 +518,18 @@ def __init__(
...
         # Create a local channel for the writer and the local readers.
         num_local_readers = len(self._reader_and_node_list) - len(
             remote_reader_and_node_list
         )
         if num_local_readers > 0:
-            local_channel = IntraProcessChannel(num_local_readers)
+            # Use num_readers = 1 when creating the local channel,
+            # because we have channel cache to support reading
+            # from the same channel multiple times.
+            local_channel = IntraProcessChannel(num_readers=1)
```
Comment (stephanie-wang) [inline python/ray/experimental/channel/shared_memory_channel.py:532]: Seems we can either remove IntraProcessChannel or make its read/write ops a no-op?

File: python/ray/dag/tests/experimental/test_accelerated_dag.py (hunk 6 above)
Comment (stephanie-wang) [inline python/ray/dag/tests/experimental/test_accelerated_dag.py:270]: For tests of the new behavior, can you also check behavior when we pass different values to `dag.execute`? I.e. check that different executions are picking up the correct cached values.

Comment (stephanie-wang) [inline python/ray/dag/tests/experimental/test_accelerated_dag.py:270]: Also add tests for:
- same task reads one input twice
- check that serialization context is empty at the end of a DAG execution (or ideally once the last task to read an input has started)

**Comment by stephanie-wang** [comment]
> > One suggestion might be to have an args cache per task, instead of a cache per channel. Then we can also do a list lookup instead of dict.
> 
> Why is a list lookup preferred than a dict? It is performance concerns (hash function overhead)?
> 
> > The first time we deserialize an arg, we put it into the inner lists for all reader tasks.
> 
> How do we know where to put into the inner lists? We still need a dict to maintain that info? Otherwise we need to go through all items of all inner lists and replace?

Yes, I thought list lookup would be better for performance and I think it also makes the garbage collection simpler. But the new approach here also looks OK.

**Comment by ruisearch42** [comment]
Still WIP with updated approach, will clean up more!

**Comment by rkooo567** [comment]
@ruisearch42 plz remove author-action-required when it is ready!

File: python/ray/experimental/channel/shared_memory_channel.py (hunk 19 above)
Comment (ruisearch42) [reply python/ray/experimental/channel/shared_memory_channel.py:532]: My plan is to remove `IntraProcessChannel` and use `CachedChannel` instead.

File: python/ray/dag/tests/experimental/test_accelerated_dag.py (hunk 6 above)
Comment (ruisearch42) [reply python/ray/dag/tests/experimental/test_accelerated_dag.py]: Done.

File: python/ray/experimental/channel/cached_channel.py (hunk 16 above)
Comment (ruisearch42) [reply python/ray/experimental/channel/cached_channel.py]: The data is auto cleaned in `get_data()` when expected number of reads are done.
Added a test in `test_channel`.

File: python/ray/dag/tests/experimental/test_accelerated_dag.py (hunk 6 above)
Comment (ruisearch42) [reply python/ray/dag/tests/experimental/test_accelerated_dag.py]: Updated the tests to run `execute()` multiple times with different value. Let me know if you suggested more than this.

File: python/ray/dag/tests/experimental/test_accelerated_dag.py (hunk 7 above)
Comment (ruisearch42) [reply python/ray/dag/tests/experimental/test_accelerated_dag.py:287]: hmm, can you clarify the example?
`fwd` is just called by s1 once, and s2 once? How many args does `fwd` take? 

Did you mean something like this?
```
    with InputNode() as inp:
        output1 = a1.inc.bind(inp)
        output2 = a1.inc.bind(inp)
        output3 = a2.inc.bind(inp)
        dag = MultiOutputNode([output1, output2, output3])
```

File: python/ray/experimental/channel/cached_channel.py (hunk 15 above)
Comment (ruisearch42) [reply python/ray/experimental/channel/cached_channel.py:17]: Renamed this to `num_reads` and clarified the doc.

File: python/ray/dag/tests/experimental/test_accelerated_dag.py [hunk 20]
```diff
@@ -1424,6 +1462,24 @@ def get_node_id(self):
     compiled_dag.teardown()
 
 
+@ray.remote
+class TestWorker:
```
Comment (ruisearch42) [inline python/ray/dag/tests/experimental/test_accelerated_dag.py:1466]: Had to move this out otherwise a weird error happens:

```
E                       ray::Replica.__init__() (pid=19780, ip=127.0.0.1, actor_id=0dee9f2dbd6be769cf7cf52d01000000, repr=<test_accelerated_dag.FunctionActorManager._create_fake_actor_class.<locals>.TemporaryActor object at 0x109b04880>)
E                       ModuleNotFoundError: No module named 'test_accelerated_dag'
```

File: python/ray/experimental/channel/shared_memory_channel.py (hunk 19 above)
Comment (ruisearch42) [reply python/ray/experimental/channel/shared_memory_channel.py:532]: `CachedChannel` now should be able to replace `IntraProcessChannel` when the `inner_channel` is `None`.
However, I plan to do it in a follow up PR for a few reasons:
* The changes are modular: this PR focuses on introducing a caching mechanism and providing new capablity; while the next PR focuses on replacement/cleanup.
* Provide the caching feature earlier to users
* Replacing `IntraProcessChannel` needs quite some changes in code and tests. It is cleaner to separate the two.

File: python/ray/tests/test_channel.py (hunk 5 above)
Comment (ruisearch42) [reply python/ray/tests/test_channel.py:925]: `IntraProcessChannel` is now only read once, and the caching is handled by `CachedChannel`. In future we will replace `IntraProcessChannel`

File: python/ray/dag/compiled_dag_node.py [hunk 21]
```diff
@@ -235,31 +240,67 @@ class _ExecutableTaskInput:
...
+            channel = adapter.get_dag_input_channel()
+
+        # Get channel result, either from cache or from channel_results
+        # NOTE: Currently we make a contract with aDAG users that the
+        # channel results should not be mutated by the actor methods.
+        # When the user needs to modify the channel results, they should
+        # make a copy of the channel results and modify the copy.
+        # This is the same contract as used in IntraProcessChannel.
+        # This contract is NOT enforced right now in either case.
+        # TODO(rui): introduce a flag to control the behavior:
```
[the remark this answers is not in the document]
Comment (ruisearch42) [reply python/ray/dag/compiled_dag_node.py]: Done: https://github.com/ray-project/ray/issues/47409

**Comment by ruisearch42** [comment]
Looking into some GPU test CI failures, but PR is ready for another look.

File: python/ray/dag/compiled_dag_node.py [hunk 22]
```diff
@@ -1110,34 +1116,67 @@ def _get_or_compile(
 
         # Create executable tasks for each actor
         for actor_handle, tasks in self.actor_to_tasks.items():
-            executable_tasks = []
+            # Dict from a channel read by the actor to the number
+            # of times it is read.
+            num_channel_reads: {ChannelInterface, int} = defaultdict(int)
```
Comment (stephanie-wang) [inline python/ray/dag/compiled_dag_node.py:1121]:
```suggestion
            num_channel_reads: Dict[ChannelInterface, int] = defaultdict(int)
```

File: python/ray/dag/compiled_dag_node.py [hunk 23]
```diff
@@ -1110,34 +1116,67 @@ def _get_or_compile(
...
                         "ray.dag.InputNode or at least one other DAGNode as an "
                         "input"
                     )
+
+            # Step 2: create cached channels if needed
+
+            # Dict from original channel to the channel to be used in execution.
+            # The value of this dict is either the original channel or a newly
+            # created CachedChannel (if the original channel is read more than once).
+            channel_dict: {ChannelInterface, ChannelInterface} = {}
```
Comment (stephanie-wang) [inline python/ray/dag/compiled_dag_node.py:1152]:
```suggestion
            channel_dict: Dict[ChannelInterface, ChannelInterface] = {}
```

File: python/ray/experimental/channel/cached_channel.py [hunk 24]
```diff
@@ -0,0 +1,109 @@
...
+    `num_reads` reads have completed. If inner channel is None, the data
+    is written to serialization context and retrieved from there. This is useful
+    when passing data within the same actor and a shared memory channel can be
+    avoided.
+
+    Args:
+        num_reads: The number of reads from this channel that must happen before
+            writing again. Readers must be methods of the same actor.
+        inner_channel: The inner channel to cache data from. If None, the data is
+            read from the serialization context.
```
Comment (stephanie-wang) [inline python/ray/experimental/channel/cached_channel.py:19]: Also document `_channel_id`?

File: python/ray/experimental/channel/cached_channel.py [hunk 25]
```diff
@@ -0,0 +1,109 @@
...
+            return ctx.get_data(self._channel_id)
+
+        assert (
+            self._inner_channel is not None
+        ), "Cannot read from the serialization context while inner channel is None."
+        value = self._inner_channel.read(timeout)
+        # The data is already read once from the inner channel,
+        # the rest of reads will be from the channel context.
+        num_reads = self._num_reads - 1
+        if num_reads > 0:
```
Comment (stephanie-wang) [inline python/ray/experimental/channel/cached_channel.py:89]: Nit: this logic is a bit easier to read if you switch the order:

```
If data not in cache:
  read from channel
  set data in cache
get from cache
```

File: python/ray/experimental/channel/shared_memory_channel.py (hunk 19 above)
Comment (stephanie-wang) [reply python/ray/experimental/channel/shared_memory_channel.py:532]: Hmm okay, sounds good. Btw, I think modifying IntraProcessChannel to have a no-op read might be a bit cleaner because then `CachedChannel` can always take a `ChannelInterface` instead of branching on whether one is provided or not. But I don't have a strong preference.

**Review by stephanie-wang — APPROVED** [review]

File: python/ray/experimental/channel/shared_memory_channel.py (hunk 19 above)
Comment (ruisearch42) [reply python/ray/experimental/channel/shared_memory_channel.py:532]: Sounds like a good idea! I will probably take that approach. 

Will do it afterwards since it will be a backwards incompatible change and will break `IntraProcessChannel` tests.

File: python/ray/dag/tests/experimental/test_accelerated_dag.py (hunk 6 above)
Comment (ruisearch42) [reply python/ray/dag/tests/experimental/test_accelerated_dag.py]: Done.

File: python/ray/dag/compiled_dag_node.py [hunk 26]
```diff
@@ -1110,34 +1116,93 @@ def _get_or_compile(
...
+                        input_channel = channel_dict[self.dag_input_channel]
+                        input_adapter = DAGInputAdapter(arg, input_channel)
+                        resolved_args.append(input_adapter)
+                    elif isinstance(arg, DAGNode):  # Other DAGNodes
+                        arg_idx = self.dag_node_to_idx[arg]
+                        arg_channel = self.idx_to_task[arg_idx].output_channel
+                        assert arg_channel is not None
+                        arg_channel = channel_dict[arg_channel]
+                        resolved_args.append(arg_channel)
+                    else:
```
Comment (rkooo567) [inline python/ray/dag/compiled_dag_node.py:1204]: i know it is just a refactoring but can you add comment what this case is?

**Review by rkooo567 — APPROVED** [review]
Looks very clean! Can you run a microbenchmark and address last test comments? let's merge it after that

File: python/ray/dag/tests/experimental/test_accelerated_dag.py [hunk 27]
```diff
@@ -267,6 +267,44 @@ def test_actor_method_bind_same_constant(ray_start_regular):
     compiled_dag.teardown()
 
 
+def test_actor_method_bind_same_input(ray_start_regular):
```
[the remark this answers is not in the document]
Comment (ruisearch42) [reply python/ray/dag/tests/experimental/test_accelerated_dag.py:270]: Good catch. Added a unit test.

File: python/ray/dag/tests/experimental/test_accelerated_dag.py [hunk 28]
```diff
@@ -267,6 +267,44 @@ def test_actor_method_bind_same_constant(ray_start_regular):
...
+    actor = Actor.remote(0)
+    with InputNode() as inp:
+        # Test binding input node to the same method
+        # of same actor multiple times: execution
+        # should not hang.
+        output1 = actor.inc.bind(inp)
```
[the remark this answers is not in the document]
Comment (ruisearch42) [reply python/ray/dag/tests/experimental/test_accelerated_dag.py:276]: Added a unit test.

File: python/ray/dag/compiled_dag_node.py (hunk 26 above)
Comment (ruisearch42) [reply python/ray/dag/compiled_dag_node.py:1204]: Added a comment, this is for constant arg.

**Comment by ruisearch42** [comment]
Kicked off microbenchmark: https://buildkite.com/ray-project/release/builds/21924#0191a221-42bb-4bd7-8f01-25adf035540e

**Comment by rkooo567** [comment]
@ruisearch42 plz comment if there's any change here! if not, let's just merge it

File: python/ray/experimental/channel/cached_channel.py [hunk 29]
```diff
@@ -0,0 +1,107 @@
...
+
+    def __str__(self) -> str:
+        return (
+            f"CachedChannel(channel_id={self._channel_id}, "
+            f"num_reads={self._num_reads}), "
+            f"inner_channel={self._inner_channel})"
+        )
+
+    def write(self, value: Any, timeout: Optional[float] = None):
+        from ray.experimental.channel import ChannelContext
```
[the remark this answers is not in the document]
Comment (rkooo567) [reply python/ray/experimental/channel/cached_channel.py:61]: hmm that's kind of sketchy. can you add TODO to address this?

**Comment by ruisearch42** [comment]
@rkooo567 microbenchmark aligns with past runs:
https://buildkite.com/ray-project/release/builds/21950#0191a40e-897c-44b1-b1f9-388ea3761ff6

The latest code just added a TODO comment compared to the last version which passed full CI.

**Comment by rkooo567** [comment]
@ruisearch42 auto merged enabled. so we should follow up with removing intra process channel right?

**Comment by rkooo567** [comment]
premerge failure. maybe consider to merge latest master?
